### PR TITLE
Initial implementation of incremental unit testing.

### PR DESCRIPTION
**Pull request by janciesko**
Initial implementation for this feature. Statically assumes one test in ./core/unit-test/incremental for now.

### CODE REVIEW

**Comment by masterleinad** [comment]
Add to whitelist

**Comment by jeffmiles63** [comment]
Refer to issue #2424 
We still have a few additions to make before it is complete.  Just opened up for discussion.

File: core/unit_test/CMakeLists.txt [hunk 1]
```diff
@@ -293,5 +294,61 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
...
+  ENDFOREACH()
+ENDFUNCTION()
+
+
+FUNCTION (KOKKOS_ADD_INCREMENTAL_TEST BACK_END_NAME)
+
+  # Iterate over incremental tests in directory
+    # Call here the bash script to get the file list in the test directory
+    # OUTPUT: List of test file names
+  SET( INCREMENTAL_FILE_LIST ExecSpaceCompilation.hpp )
```
Comment (jeffmiles63) [inline core/unit_test/CMakeLists.txt]: Note that this is a place holder...variable INCREMENTAL_FILE_LIST will be populated with script above

File: core/unit_test/CMakeLists.txt [hunk 2]
```diff
@@ -293,5 +294,61 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
...
+  SET( ${BACK_END_NAME}_SOURCES UnitTestMainInit.cpp )
+
+  foreach (CURRENT_FILE_NAME ${INCREMENTAL_FILE_LIST})
+    string (REPLACE ".hpp" "" CURRENT_TEST_NAME ${CURRENT_FILE_NAME})
+
+    message( STATUS "OUT:${CURRENT_TEST_NAME}" )
+    # From each test get list of dependencies 
+      # Run a bash script to get the list of dependencies for this test
+      # OUTPUT: List of Kokkos features required to run the test
+    SET( CURRENT_REQUIRED_FEATURE_LIST "ExecSpaceDef" )
```
Comment (jeffmiles63) [inline core/unit_test/CMakeLists.txt]: note that this is a place holder...variable CURRENT_REQUIRED_FEATURE_LIST will be populated with script above

File: core/unit_test/CMakeLists.txt [hunk 3]
```diff
@@ -293,5 +294,61 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
...
+    ${${BACK_END_NAME}_SOURCES}
+  )
+
+  #SET(EXE_NAME ${PACKAGE_NAME}_${ROOT_NAME})
+  TARGET_INCLUDE_DIRECTORIES( ${PACKAGE_NAME}_IncrementalTest_${BACK_END_NAME} PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/incremental )
 
+ENDFUNCTION()
 
+#KOKKOS_ADD_INCREMENTAL_TEST ( OpenMPTarget )
+KOKKOS_ADD_INCREMENTAL_TEST ( OpenMP )
```
Comment (jeffmiles63) [inline core/unit_test/CMakeLists.txt]: We had to pick something to test with...still need a way to select in-progress experimental backends

File: core/unit_test/IncrementalTest.cpp.in [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+*/
+
+#ifndef KOKKOS_TEST_@BACK_END_NAME@
+#define KOKKOS_TEST_@BACK_END_NAME@
+
+#include <gtest/gtest.h>
+
+#define TEST_CATEGORY @BACK_END_NAME@
+//#define TEST_EXECSPACE Kokkos::Experimental::@BACK_END_NAME@
+#define TEST_EXECSPACE Kokkos::@BACK_END_NAME@
```
Comment (jeffmiles63) [inline core/unit_test/IncrementalTest.cpp.in]: This is only because of the OpenMP reference above

File: core/src/Kokkos_Core.hpp [hunk 5]
```diff
@@ -57,10 +57,10 @@
 #include <Kokkos_OpenMP.hpp>
 #endif
 
-//#if defined( KOKKOS_ENABLE_OPENMPTARGET )
```
Comment (jeffmiles63) [inline core/src/Kokkos_Core.hpp]: Changes in this file aren't necessary for this.

File: cmake/KokkosCore_config.h.in [hunk 6]
```diff
@@ -8,6 +8,7 @@
 /* Execution Spaces */
 #cmakedefine KOKKOS_ENABLE_SERIAL
 #cmakedefine KOKKOS_ENABLE_OPENMP
+#cmakedefine KOKKOS_ENABLE_OPENMPTARGET
```
Comment (jeffmiles63) [inline cmake/KokkosCore_config.h.in]: Changes in this file shouldn't be necessary.

File: core/unit_test/CMakeLists.txt (hunk 1 above)
Comment (crtrott) [reply core/unit_test/CMakeLists.txt]: I don't think we need a bash script to build the file list can't we just do this in cmake?

File: core/unit_test/CMakeLists.txt (hunk 2 above)
Comment (crtrott) [reply core/unit_test/CMakeLists.txt]: I have the feeling while potentially powerful, having a full dependency list of features is more than we need. I would be fairly comfortable with just numbering the tests through, and have for each execution space a max number up to which stuff is implemented, and maybe a list of "to-be-skipped" tests.

File: core/unit_test/CMakeLists.txt (hunk 1 above)
Comment (jeffmiles63) [reply core/unit_test/CMakeLists.txt]: Switched to use GLOB for file list in incremental

File: core/unit_test/CMakeLists.txt (hunk 2 above)
Comment (DavidPoliakoff) [reply core/unit_test/CMakeLists.txt]: @Rombur said the move to a feature set was necessary, that there isn't a strict ordering of tests. Bruno, would a capability number plus a list of tests to skip meet your needs?

Comment (Rombur) [reply core/unit_test/CMakeLists.txt]: Yes, that would work.

Comment (jeffmiles63) [reply core/unit_test/CMakeLists.txt]: updated this to use a required level and the -DKokkos_<>_FEATURE_LEVEL must be greater than or equal to required level in order for the test to be included.

Comment (jeffmiles63) [reply core/unit_test/CMakeLists.txt]: Also added test exclusion logic

File: core/unit_test/CMakeLists.txt [hunk 7]
```diff
@@ -293,5 +294,61 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
...
+  
+  # Iterate over incremental tests in directory
+  APPEND_GLOB(INCREMENTAL_FILE_LIST ${CMAKE_CURRENT_SOURCE_DIR}/incremental/*.hpp)
+  foreach (CURRENT_FILE_PATH ${INCREMENTAL_FILE_LIST})
+    get_filename_component( CURRENT_FILE_NAME ${CURRENT_FILE_PATH} NAME )
+    string (REPLACE ".hpp" "" CURRENT_TEST_NAME ${CURRENT_FILE_NAME})
+    message( STATUS "OUT:${CURRENT_TEST_NAME}" )
+    SET ( TEST_IMPLEMENTED ON )
+    if (Kokkos_${BACK_END_NAME}_EXCLUDE_TESTS)
+       TEST_EXCLUDED( ${CURRENT_TEST_NAME} ${Kokkos_${BACK_END_NAME}_EXCLUDE_TESTS} )
```
Comment (dalg24) [inline core/unit_test/CMakeLists.txt]: You don't need a function for that.  Just do
```cmake
if(NOT CURRENT_TEST_NAME IN_LIST Kokkos_${BACK_END_NAME}_EXCLUDE_TESTS)
  ...
endif()
```

File: core/unit_test/incremental/ExecSpaceCompilation.hpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+template <class ExecSpace>
+struct TestIncrExecSpaceTypedef {
+
+  void testit() {
+    bool passed = true;
+    passed &= !std::is_same<void,typename ExecSpace::memory_space>::value;
+    passed &=  std::is_same<ExecSpace,typename ExecSpace::execution_space>::value;
+    passed &= !std::is_same<void,typename ExecSpace::scratch_memory_space>::value;
+    passed &= !std::is_same<void,typename ExecSpace::array_layout>::value;
+    ASSERT_TRUE(passed);
```
Comment (dalg24) [inline core/unit_test/incremental/ExecSpaceCompilation.hpp]: Please replace this with a series of assertions.

Comment (dalg24) [reply core/unit_test/incremental/ExecSpaceCompilation.hpp]: Also I am temped to make these static assertions.

Comment (dalg24) [reply core/unit_test/incremental/ExecSpaceCompilation.hpp]: (You can ignore the static assertion comment if you don't agree.  I expect you to address all others comments)

File: core/unit_test/CMakeLists.txt (hunk 7 above)
Comment (jeffmiles63) [reply core/unit_test/CMakeLists.txt]: Fixed this..
Also added Kokkos_${BACK_END_NAME}_EXCLUDE_TESTS and Kokkos_${BACK_END_NAME}_FEATURE_LEVEL to CMakeCache using KOKKOS_OPTION

File: core/unit_test/CMakeLists.txt [hunk 9]
```diff
@@ -293,5 +294,61 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
   SOURCES UnitTestMain.cpp  TestHostBarrier.cpp
 )
 
+FUNCTION (TEST_EXCLUDED CURRENT_TEST TEST_EXCLUDE_LIST)
+  FOREACH ( EXCLUDED_TEST ${TEST_EXCLUDE_LIST})
+    if(CURRENT_TEST STREQUAL ${EXCLUDED_TEST} )
+      SET (TEST_IMPLEMENTED OFF PARENT_SCOPE )
+    endif()
+  ENDFOREACH()
```
[the remark this answers is not in the document]
Comment (jeffmiles63) [reply core/unit_test/CMakeLists.txt]: Fixed this

File: core/unit_test/incremental/ExecSpaceCompilation.hpp (hunk 8 above)
Comment (crtrott) [reply core/unit_test/incremental/ExecSpaceCompilation.hpp]: Yeah make it static_asserts.

File: core/unit_test/CMakeLists.txt (hunk 2 above)
Comment (crtrott) [reply core/unit_test/CMakeLists.txt]: This looks much better to me. Thanks.

File: core/unit_test/CMakeLists.txt [hunk 10]
```diff
@@ -293,5 +294,46 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
...
 )
 
+FUNCTION (KOKKOS_ADD_INCREMENTAL_TEST BACK_END_NAME)
+  KOKKOS_OPTION( ${BACK_END_NAME}_EXCLUDE_TESTS "" STRING "Incremental test exclude list" )
+  KOKKOS_OPTION( ${BACK_END_NAME}_FEATURE_LEVEL "0" STRING "Incremental test implementation level" )
+  # Add unit test main
+  SET(${BACK_END_NAME}_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/UnitTestMainInit.cpp)
+
+  # Iterate over incremental tests in directory
+  APPEND_GLOB(INCREMENTAL_FILE_LIST ${CMAKE_CURRENT_SOURCE_DIR}/incremental/*.hpp)
```
Comment (jeffmiles63) [inline core/unit_test/CMakeLists.txt]: It just occurred to me that this (INCREMENTAL_FILE_LIST) only needs to be populated once...we should move this to before we process the experimental backend inputs and call KOKKOS_ADD_INCREMENTAL_TEST the first time.

File: core/unit_test/CMakeLists.txt [hunk 11]
```diff
@@ -293,5 +294,46 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
...
+    IncrementalTest_${BACK_END_NAME}
+    SOURCES
+    ${${BACK_END_NAME}_SOURCES}
+  )
+
+  TARGET_INCLUDE_DIRECTORIES( ${PACKAGE_NAME}_IncrementalTest_${BACK_END_NAME} PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/incremental )
 
+ENDFUNCTION()
 
+FOREACH (BACK_END_NAME ${Kokkos_INCREMENTAL_TEST_BACK_ENDS})
```
Comment (jeffmiles63) [inline core/unit_test/CMakeLists.txt]: See comment above about INCREMENTAL_FILE_LIST.  We should populate the file list prior to this loop

**Comment by rgayatri23** [comment]
I addressed all the comments made for this pull request 
1) cmake changes - updated by jan
2) Replaced asserts with static assertions
3) Formatted the hpp file with clang 8 format.

File: core/unit_test/incremental/01_execspace.hpp [hunk 12]
```diff
@@ -0,0 +1,97 @@
+/*
```
Comment (jeffmiles63) [inline core/unit_test/incremental/01_execspace.hpp]: we should remove the header files from this PR before merging.  The header files should be in subsequent PRs #2536, #2537, #2538, #2539

**Comment by janciesko** [comment]
PR is pointing to the incremental-testing branch. I have tested with two incremental unit tests and for different back-ends at different simulated implementation levels. Feel free to accept.

File: core/unit_test/incremental/01_execspace.hpp (hunk 12 above)
Comment (janciesko) [reply core/unit_test/incremental/01_execspace.hpp]: Done.

File: core/unit_test/incremental/ExecSpaceCompilation.hpp (hunk 8 above)
Comment (jeffmiles63) [reply core/unit_test/incremental/ExecSpaceCompilation.hpp]: we removed the actual tests from this PR

File: core/unit_test/IncrementalTest.cpp.in (hunk 4 above)
Comment (jeffmiles63) [reply core/unit_test/IncrementalTest.cpp.in]: removed

**Comment by rgayatri23** [comment]
Applied the clang-8 format to the file.

**Pull request merged**